v2/yarpcgrpc/dialer.go

```diff
@@ -21,30 +21,82 @@
 package grpc
 
 import (
-	"go.uber.org/yarpc/api/peer"
+	"context"
+	"sync"
+
+	"go.uber.org/multierr"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpcpeer"
 )
 
-// NewDialer creates a transport that is decorated to retain peers with
-// additional gRPC dial options.
-func (t *Transport) NewDialer(options ...DialOption) *Dialer {
-	return &Dialer{trans: t, options: newDialOptions(options)}
-}
+var _ yarpc.Dialer = (*Dialer)(nil)
 
 // Dialer is a decorator for a gRPC transport that threads dial options for
 // every retained peer.
 type Dialer struct {
-	trans   *Transport
-	options *dialOptions
+	lock          sync.Mutex
+	options       *dialOptions
+	addressToPeer map[string]*grpcPeer
 }
 
-var _ peer.Transport = (*Dialer)(nil)
+// Start starts the gRPC dialer.
+func (d *Dialer) Start(_ context.Context) error {
+	d.addressToPeer = make(map[string]*grpcPeer)
+	// init options?
+	return nil
+}
+
+// Stop stops the gRPC dialer.
+func (d *Dialer) Stop() error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for _, grpcPeer := range d.addressToPeer {
+		grpcPeer.stop()
+	}
+	var err error
+	for _, grpcPeer := range d.addressToPeer {
+		err = multierr.Append(err, grpcPeer.wait())
+	}
+	return err
+}
 
 // RetainPeer retains the identified peer, passing dial options.
-func (d *Dialer) RetainPeer(id peer.Identifier, ps peer.Subscriber) (peer.Peer, error) {
-	return d.trans.retainPeer(id, d.options, ps)
+func (d *Dialer) RetainPeer(id yarpc.Identifier, ps yarpc.Subscriber) (yarpc.Peer, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	address := id.Identifier()
+	p, ok := d.addressToPeer[address]
+	if !ok {
+		var err error
+		p, err = d.newPeer(address, d.options)
+		if err != nil {
+			return nil, err
+		}
+		d.addressToPeer[address] = p
+	}
+	p.Subscribe(ps)
+	return p, nil
 }
 
 // ReleasePeer releases the identified peer.
-func (d *Dialer) ReleasePeer(id peer.Identifier, ps peer.Subscriber) error {
-	return d.trans.ReleasePeer(id, ps)
+func (d *Dialer) ReleasePeer(id yarpc.Identifier, ps yarpc.Subscriber) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	address := id.Identifier()
+	p, ok := d.addressToPeer[address]
```

Comment:
Given that we've seen this form a couple times, we could make a helper that takes advantage of the read locking behavior (mentioned above):
```
func (d *Dialer) getPeer(address string) (peer.Peer, bool) {
  d.lock.RLock()
  p, ok := d.addressToPeer(address)
  d.lock.RUnlock()
  return p, ok
}
```

Reply:
similar to above, I'd rather keep this as one atomic event for now. I'm open to more optimizations in a follow up since this is already quite large.